### PR TITLE
Enh #7405: Added `yii\filters\auth\AuthMetod::optional` property

### LINKED ISSUE #7405: Optional authentication in REST
I've faced situation when single action should be available to guest users and to authenticated users.
What is a best way to solve it?
First idea - implement new attribute `optional`:
```
'authenticator' => [
    'class' => CompositeAuth::className(),
    'optional' => ['feed', 'feed-item'],
    'authMethods' => [
        HttpBearerAuth::className(),
    ],
],
```
Methods `actionFeed` and `actionFeedItem` will perform Bearer authentication if header is passed. Otherwise authentication should be ignored and no errors should be raised.
Thats needed because application's pages for guest & authenticated users are nearly same, excluding a bit of additional data for authenticated guys.

### PR DESCRIPTION
**Pull request by SilverFire**
Closes #7405
- [x] Code
- [x] Tests
- [x] Changelog
- [x] Code review

### CODE REVIEW

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ Yii Framework 2 Change Log
 - Bug #9791: Fixed endless loop on file creation for non-existing device letters on windows (lukos, cebe)
 - Bug #9874: Fixed outputting exception stacktrace in non-debug mode when `Response::FORMAT_RAW` is used (nainoon)
 - Bug #9883: Passing a single `yii\db\Expression` to `Query::select()` or `::addSelect()` was not handled correctly in all cases (cebe)
-- Bug #9911: Fixed `yii\helpers\BaseStringHelper::explode()` code so it does not remove items eq to 0 with skip_empty attribute (silverfire, kidol)
+- Bug #9911: Fixed `yii\helpers\BaseStringHelper::explode()` code so it does not remove items eq to 0 with `skip_empty` attribute (silverfire, kidol)
```
Comment (samdark) [inline framework/CHANGELOG.md]: eq → equal

Comment (samdark) [inline framework/CHANGELOG.md]: with `skip_empty` attribute  → when`skip_empty` attribute is true

Comment (samdark) [inline framework/CHANGELOG.md]: Not about this issue but still :)

File: framework/base/ActionFilter.php [hunk 2]
```diff
@@ -127,6 +124,18 @@ protected function isActive($action)
         } else {
             $id = $action->id;
         }
+
+        return $id;
+    }
+
+    /**
+     * Returns a value indicating whether the filter is active for the given action.
```
Comment (samdark) [inline framework/base/ActionFilter.php]: It's a good idea to add `@since 2.0.7`.

File: framework/base/ActionFilter.php [hunk 3]
```diff
@@ -111,14 +110,12 @@ public function afterAction($action, $result)
...
-     * Returns a value indicating whether the filter is active for the given action.
-     * @param Action $action the action being filtered
-     * @return boolean whether the filter is active for the given action.
+     * Returns an $action ID, convert action uniqueId into an ID relative to the module
+     * @param Action $action
+     * @return string
      */
-    protected function isActive($action)
-    {
+    protected function getActionId($action) {
```
Comment (samdark) [inline framework/base/ActionFilter.php]: It's a good idea to add `@since 2.0.7`.

File: framework/base/ActionFilter.php (hunk 2 above)
Comment (SilverFire) [inline framework/base/ActionFilter.php]: It's not since 2.0.7. Part of the method was extracted to `getActionId`

**Comment by samdark** [comment]
Looks good to me.

**Comment by SilverFire** [comment]
Merged https://github.com/yiisoft/yii2/commit/035a9ce0a8642dc1218f4b5d587dfde4daa382f2

**Pull request closed without merging**